zkevm-circuits/src/witness/tx.rs

```diff
@@ -389,12 +389,40 @@ impl Transaction {
             byte_idx: 0,
             depth: 0,
         };
+
+        // Queue up stack operations
+        let mut stack_ops: Vec<RlpStackOp<F>> = vec![];
+        let mut last_bytes_on_depth: [usize; 4] = [0, 0, 0, 0];
+        let mut stack_acc = Value::known(F::zero());
+
+        let mut stack_acc_pow_of_rand: Vec<Value<F>> = vec![];
+        let mut pow_of_rand = Value::known(F::one());
+        stack_acc_pow_of_rand.push(pow_of_rand);
+        for _ in 0..4 {
+            pow_of_rand = pow_of_rand * keccak_rand;
+            stack_acc_pow_of_rand.push(pow_of_rand);
+        }
+        // concat tx_id and format as stack identifier
+        let id = keccak_rand * Value::known(F::from(tx_id)) + Value::known(F::from(format as u64));
         // When we are decoding a vector of element type `t`, at the beginning
         // we actually do not know the next tag is `EndVector` or not. After we
         // parsed the current tag, if the remaining bytes to decode in this layer
         // is zero, then the next tag is `EndVector`.
         let mut cur_rom_row = vec![0];
         let mut remaining_bytes = vec![rlp_bytes.len()];
+        // initialize stack
+        stack_ops.push(RlpStackOp {
+            is_write: true,
+            id,
+            depth: cur.depth,
+            value: rlp_bytes.len(),
+            value_prev: 0,
+            stack_acc: Value::known(F::zero()),
+            is_init: true,
+            is_push: false,
+            is_pop: false,
+            is_update: false,
```

Comment:
I would suggest a refactor here, to something like:

```suggestion
        stack_ops.push(RlpStackOp::init(/* args */));
```

And similarly in other places, we can have methods `RlpStackOp::push`, `RlpStackOp::update` and `RlpStackOp::pop`.

It makes the code more readable, and easier to debug in case a particular field was incorrectly set.

Reply:
Refactored. Additional constructors for `RlpStackOp` now include `init`, `push`, `pop`, and `update`.